pydantic/_internal/_generate_schema.py

```diff
@@ -210,6 +211,42 @@ def modify_model_json_schema(
     return json_schema
 
 
+JsonEncoders = Dict[Type[Any], JsonEncoder]
+
+
+def _add_custom_serialization_from_json_encoders(
+    json_encoders: JsonEncoders | None, tp: Any, schema: CoreSchema
+) -> CoreSchema:
+    """Iterate over the json_encoders and add the first matching encoder to the schema.
+
+    Args:
+        json_encoders (JsonEncoders | None): A dictionary of types and their encoder functions.
+        tp (Any): The type to check for a matching encoder.
+        schema (CoreSchema): The schema to add the encoder to.
+    """
+    if json_encoders is None:
+        return schema
+    if 'serialization' in schema:
+        return schema
+    # Check the class type and its superclasses for a matching encoder
+    # Decimal.__class__.__mro__ (and probably other cases) doesn't include Decimal itself
+    for base in (tp, *tp.__class__.__mro__[:-1]):
+        try:
+            encoder = json_encoders[base]
+        except KeyError:
+            continue
+
+        serialization = schema.get('serialization') or core_schema.plain_serializer_function_ser_schema(encoder)
```

Comment:
```suggestion
        serialization = core_schema.plain_serializer_function_ser_schema(encoder)
```
It already returns above if `schema.get('serialization') is not None`, so I think this or can be dropped